include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,199 @@
+                          );
+    }
+
+    soa_storage(soa_storage const& src) : soa_storage(src.storage_.get_deleter())
+    {
+      auto sz = byte_size(src.capacity_);
+
+      // Strong guarantee on allocation
+      auto local  = memory_helper::allocate(*this,sz);
+      memory_helper::copy_all_aligned(src.storage_.get(), local.get(), sz );
+      storage_.swap(local);
+
+      indexes_  = src.indexes_;
+      capacity_ = src.capacity_;
+    }
+
+    soa_storage& operator=(soa_storage const& src)
+    {
+      *this = soa_storage(src);
+      return *this;
+    }
+
+    soa_storage(soa_storage&&)            = default;
+    soa_storage& operator=(soa_storage&&) = default;
+
+    EVE_FORCEINLINE auto data_aligned()
+    {
+      auto ptrs = kumi::map([]( auto p) { return memory_helper::as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([]( auto p) { return memory_helper::as_aligned_pointer( p ); }, *this);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      storage_.swap(other.storage_);
+      std::swap(indexes_  , other.indexes_  );
+      std::swap(capacity_ , other.capacity_ );
+    }
+
+    friend void swap(soa_storage& a, soa_storage& b ) noexcept { a.swap(b); }
+
+    //==============================================================================================
+    // Static helpers
+    //==============================================================================================
+    static auto aligned_capacity(std::size_t n) noexcept
+    {
+      // Capacity is the smallest gap possible, ie the one in the largest type
+      constexpr std::size_t largest = max_scalar_size_v<Type>;
+      return eve::align(largest*n, eve::over{64})/largest;
+    }
+
+    static auto byte_size(std::size_t n) noexcept
+    {
+      return  kumi::fold_left ( [n]<typename S>(auto a,S) { return a + memory_helper::realign<S>(n); }
+                              , flat_t{}
+                              , 0ULL
+                              );
+    }
+
+    //==============================================================================================
+    // Tuple interface
+    //==============================================================================================
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type>*>(s.storage_.get() + get<I>(s.indexes_));
+    }
+
+    template<std::size_t I>
+    friend decltype(auto) get(soa_storage const& s) noexcept
+    {
+      return reinterpret_cast<kumi::element_t<I,Type> const*>(s.storage_.get() + get<I>(s.indexes_));
+    }
```

Comment:
Are the get used anywhere? They seem weird.

Reply:
They are used because soa_storage is a product_type so whenever you run over it as a tuple you get the tuple of pointer so that all the rest of the code works

Reply:
Ah - there are `kumi::map(*this)` ok.